core/src/main/clojure/xtdb/operator/window.clj

```diff
@@ -0,0 +1,237 @@
+            [xtdb.vector.writer :as vw])
+  (:import (clojure.lang IPersistentMap)
+           (java.io Closeable)
+           (java.util LinkedList List Spliterator)
+           (java.util HashMap TreeSet)
+           (org.apache.arrow.memory BufferAllocator)
+           (org.apache.arrow.vector BigIntVector)
+           (xtdb ICursor)
+           (org.apache.arrow.vector.types.pojo Field)
+           (xtdb.operator.group_by IGroupMapper)
+           (xtdb.vector RelationWriter IVectorWriter)))
+
+(s/def ::window-name symbol?)
+
+;; TODO
+(s/def ::frame any?)
+;; TODO assert at least one is present
+(s/def ::partition-cols (s/coll-of ::lp/column :min-count 1))
+(s/def ::window-spec (s/keys :opt-un [::partition-cols ::order-by/order-specs ::frame]))
+(s/def ::windows (s/map-of ::window-name ::window-spec))
+
+
+(s/def ::window-agg-expr
+  (s/or :nullary (s/cat :f simple-symbol?)
+        :unary (s/cat :f simple-symbol?
+                      :from-column ::lp/column)
+        :binary (s/cat :f simple-symbol?
+                       :left-column ::lp/column
+                       :right-column ::lp/column)))
+
+(s/def ::window-agg ::window-agg-expr)
+(s/def ::window-projection (s/map-of ::lp/column (s/keys :req-un [::window-name ::window-agg])))
+(s/def ::projections (s/coll-of ::window-projection :min-count 1))
+
+(defmethod lp/ra-expr :window [_]
+  (s/cat :op #{:window}
+         :specs (s/keys :req-un [::windows ::projections])
+         :relation ::lp/ra-expression))
+
+(comment
+  (s/valid? ::lp/logical-plan
+            '[:window {:windows {window1 {:partition-cols [a]
+                                          :order-specs [[b]]
+                                          :frame {}}}
+                       :projections [{col-name {:window-name window1
+                                                :window-agg (row-number)}}]}
+              [:table [{}]]]))
+
+
+#_{:clj-kondo/ignore [:unused-binding :clojure-lsp/unused-public-var]}
+(definterface IWindowFnSpec
+  (^void aggregate [^org.apache.arrow.vector.IntVector groupMapping
+                    ^ints sortMapping
+                    ^xtdb.vector.RelationReader in-rel])
+  (^xtdb.vector.IVectorReader finish []))
+
+#_{:clj-kondo/ignore [:unused-binding :clojure-lsp/unused-public-var]}
+(definterface IWindowFnSpecFactory
+  (^clojure.lang.Symbol getToColumnName [])
+  (getToColumnField [])
+  (^xtdb.operator.window.IWindowFnSpec build [^org.apache.arrow.memory.BufferAllocator allocator]))
+
+#_{:clj-kondo/ignore [:unused-binding]}
+(defmulti ^xtdb.operator.window.IWindowFnSpecFactory ->window-fn-factory
+  (fn [{:keys [f from-name from-type to-name zero-row?]}]
+    (expr/normalise-fn-name f)))
+
+(defmethod ->window-fn-factory :row_number [{:keys [to-name]}]
+  (reify IWindowFnSpecFactory
+    (getToColumnName [_] to-name)
+    (getToColumnField [_] (types/col-type->field :i64))
+
+    (build [_ al]
+      (let [^BigIntVector out-vec (-> (types/col-type->field to-name :i64)
```

Comment:
I'd now use an `IVectorWriter` here, save you faffing about with `setValueCount` etc below

Reply:
But I need the vector later. Are you saying I should then convert to old style?